controllers/routingctrl/controller_test.go

```diff
@@ -539,131 +544,156 @@ func hasNoAddressAnnotations(component *unstructured.Unstructured) func(g Gomega
+		}
 
 		return nil
 	}
 }
 
 func destinationRuleExistsFor(exposedSvc *corev1.Service) func(g Gomega, ctx context.Context) error {
 	return func(g Gomega, ctx context.Context) error {
-		destinationRule := &v1beta1.DestinationRule{}
-		if errGet := envTest.Get(ctx, types.NamespacedName{
-			Name:      exposedSvc.Name + "-" + exposedSvc.Namespace,
-			Namespace: routingConfiguration.GatewayNamespace,
-		}, destinationRule); errGet != nil {
-			return errGet
+		g.Expect(exposedSvc.Spec.Ports).ToNot(BeEmpty())
+
+		for _, exposedPort := range exposedSvc.Spec.Ports {
+			destinationRule := &v1beta1.DestinationRule{}
+			if errGet := envTest.Get(ctx, types.NamespacedName{
+				Name:      exposedSvc.Name + "-" + exposedPort.Name + "-" + exposedSvc.Namespace,
+				Namespace: routingConfiguration.GatewayNamespace,
+			}, destinationRule); errGet != nil {
+				return errGet
+			}
+
+			g.Expect(destinationRule).To(
+				HaveHost(
+					exposedSvc.Name + "-" + exposedPort.Name + "-" + exposedSvc.Namespace + "." + routingConfiguration.GatewayNamespace + ".svc.cluster.local",
+				),
+			)
+			g.Expect(destinationRule.Spec.GetTrafficPolicy().GetTls().GetMode()).To(Equal(istionetworkingv1beta1.ClientTLSSettings_DISABLE))
 		}
 
-		g.Expect(destinationRule).To(
-			HaveHost(
-				exposedSvc.Name + "-" + exposedSvc.Namespace + "." + routingConfiguration.GatewayNamespace + ".svc.cluster.local",
-			),
-		)
-		g.Expect(destinationRule.Spec.GetTrafficPolicy().GetTls().GetMode()).To(Equal(istionetworkingv1beta1.ClientTLSSettings_DISABLE))
-
 		return nil
 	}
 }
 
-func ingressVirtualServiceExistsFor(exportedSvc *corev1.Service) func(g Gomega, ctx context.Context) error {
+func ingressVirtualServiceExistsFor(exposedSvc *corev1.Service) func(g Gomega, ctx context.Context) error {
 	return func(g Gomega, ctx context.Context) error {
-		routerVS := &v1beta1.VirtualService{}
-		if errGet := envTest.Get(ctx, types.NamespacedName{
-			Name:      exportedSvc.Name + "-" + exportedSvc.Namespace + "-ingress",
-			Namespace: routingConfiguration.GatewayNamespace,
-		}, routerVS); errGet != nil {
-			return errGet
+		g.Expect(exposedSvc.Spec.Ports).ToNot(BeEmpty())
+
+		for _, exposedPort := range exposedSvc.Spec.Ports {
+			routerVS := &v1beta1.VirtualService{}
+			if errGet := envTest.Get(ctx, types.NamespacedName{
+				Name:      exposedSvc.Name + "-" + exposedPort.Name + "-" + exposedSvc.Namespace + "-ingress",
+				Namespace: routingConfiguration.GatewayNamespace,
+			}, routerVS); errGet != nil {
+				return errGet
+			}
+
+			g.Expect(routerVS).To(HaveHost(exposedSvc.Name + "-" + exposedPort.Name + "-" + exposedSvc.Namespace + "." + domain))
+			g.Expect(routerVS).To(BeAttachedToGateways(routingConfiguration.IngressService))
+			g.Expect(routerVS).To(RouteToHost(exposedSvc.Name+"."+exposedSvc.Namespace+".svc.cluster.local", uint32(exposedPort.TargetPort.IntValue())))
```

Comment:
This might result in a snowball effect of test failures, where we have to go one execution after the other, but I think we can address it as an improvement. However, I cannot find anything about "soft assertions" in Gomega.

Reply:
Yeah, not happy about the pattern matching stuff on this lvl of testing either.